File: app/src/main/java/net/hearnsoft/tcm/utils/ApiConfigParser.java
package net.hearnsoft.tcm.utils;

import net.hearnsoft.tcm.infrastructure.api.test.ApiTestConfig;
import net.hearnsoft.tcm.infrastructure.api.test.ApiTestEntry;
import net.hearnsoft.tcm.infrastructure.logger.Logger;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

public class ApiConfigParser {

    public ApiTestEntry parseApiConfigs(InputStream inputStream) {
        SAXParserFactory spf = SAXParserFactory.newInstance();
        ApiConfigHandler handler;
        try {
            SAXParser parser = spf.newSAXParser();
            handler = new ApiConfigHandler();
            parser.parse(inputStream, handler);
        } catch (Exception e) {
            Logger.err("ApiConfigParser", e.getMessage());
            return null;
        }
        return new ApiTestEntry(handler.getApiTestEntry());
    }

    private static class ApiConfigHandler extends DefaultHandler {
        private final List<ApiTestConfig> apis = new ArrayList<>();
        private final StringBuilder builder = new StringBuilder();
        private ApiTestConfig apiTestConfig;

        @Override
        public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
            if (qName.equalsIgnoreCase("api")) {
                apiTestConfig = new ApiTestConfig();
            }
            builder.setLength(0);
        }

        @Override
        public void characters(char[] ch, int start, int length) throws SAXException {
            builder.append(ch, start, length);
        }

        @Override
        public void endElement(String uri, String localName, String qName) throws SAXException {
            if (apiTestConfig == null) return;

            switch (qName.toLowerCase()) {
                case "name":
                    apiTestConfig.setApiName(builder.toString());
                    break;
                case "fragment":
                    apiTestConfig.setTestFragment(builder.toString());
                    break;
                case "api":
                    apis.add(apiTestConfig);
                    apiTestConfig = null;
                    break;
            }
        }

        public List<ApiTestConfig> getApiTestEntry() {
            return new ArrayList<>(apis);
        }
    }
}

File: app/src/main/java/net/hearnsoft/tcm/infrastructure/api/ApiConfigParser.kt
package net.hearnsoft.tcm.infrastructure.api

import net.hearnsoft.tcm.infrastructure.api.test.ApiTestConfig
import net.hearnsoft.tcm.infrastructure.api.test.ApiTestEntry
import net.hearnsoft.tcm.infrastructure.logger.Logger.err
import org.xml.sax.Attributes
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler
import java.io.InputStream
import java.util.Locale
import javax.xml.parsers.SAXParserFactory

class ApiConfigParser {
    fun parseStream(inputStream: InputStream?): ApiTestEntry? {
        val spf = SAXParserFactory.newInstance()
        val handler: ApiConfigHandler
        try {
            val parser = spf.newSAXParser()
            handler = ApiConfigHandler()
            parser.parse(inputStream, handler)
        } catch (e: Exception) {
            err("ApiConfigParser", e.message!!)
            return null
        }
        return ApiTestEntry(handler.apiTestEntry)
    }

    private class ApiConfigHandler : DefaultHandler() {
        private val apis: MutableList<ApiTestConfig> = ArrayList()
        private val builder = StringBuilder()
        private var apiTestConfig: ApiTestConfig? = null

        @Throws(SAXException::class)
        override fun startElement(
            uri: String,
            localName: String,
            qName: String,
            attributes: Attributes
        ) {
            if (qName.equals("api", ignoreCase = true)) {
                apiTestConfig = ApiTestConfig()
            }
            builder.setLength(0)
        }

        @Throws(SAXException::class)
        override fun characters(ch: CharArray, start: Int, length: Int) {
            builder.append(ch, start, length)
        }

        @Throws(SAXException::class)
        override fun endElement(uri: String, localName: String, qName: String) {
            if (apiTestConfig == null) return

            when (qName.lowercase(Locale.getDefault())) {
                "name" -> apiTestConfig!!.apiName = builder.toString()
                "fragment" -> apiTestConfig!!.testFragment = builder.toString()
                "api" -> {
                    apis.add(apiTestConfig!!)
                    apiTestConfig = null
                }
            }
        }

        val apiTestEntry: List<ApiTestConfig>
            get() = ArrayList(apis)
    }
}
